### PR TITLE
Implement content models

### PR DESCRIPTION
**Pull request by dan-f**
@jimabramson please review

@tlindaliu @BenjiLee fyi

### CODE REVIEW

File: discussions/models.py [hunk 1]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+
+class CommentThread(Content):
+    """Represents a Thread."""
+    meta = {'collection': 'contents'}
+    _type = StringField(default='CommentThread')
+    thread_type = StringField(default='discussion', choices=['question', 'discussion'])
+    comment_count = IntField(default=0)
+    title = StringField()
+    commentable_id = StringField()
+    closed = BooleanField(default=False)
```
Comment (BenjiLee) [inline discussions/models.py:85]: String as a boolean

File: discussions/models.py [hunk 2]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+        point = IntField(default=0)
+
+    visible = BooleanField(default=True)
+    abuse_flaggers = ListField()
+    historical_abuse_flaggers = ListField()
+    author_username = StringField(default=None)
+    course_id = StringField()
+    body = StringField()
+    anonymous = BooleanField(default=False)
+    anonymous_to_peers = BooleanField(default=False)
```
Comment (BenjiLee) [inline discussions/models.py:69]: String as a boolean

File: discussions/models.py [hunk 3]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+        count = IntField(default=0)
+        point = IntField(default=0)
+
+    visible = BooleanField(default=True)
+    abuse_flaggers = ListField()
+    historical_abuse_flaggers = ListField()
+    author_username = StringField(default=None)
+    course_id = StringField()
+    body = StringField()
+    anonymous = BooleanField(default=False)
```
Comment (BenjiLee) [inline discussions/models.py]: String as a boolean

File: discussions/models.py [hunk 4]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+        up = ListField()
+        down = ListField()
+        up_count = IntField(default=0)
+        down_count = IntField(default=0)
+        count = IntField(default=0)
+        point = IntField(default=0)
+
+    visible = BooleanField(default=True)
+    abuse_flaggers = ListField()
+    historical_abuse_flaggers = ListField()
```
Comment (jimabramson) [inline discussions/models.py]: `default=[]` i think

Comment (jimabramson) [inline discussions/models.py]: for both

File: discussions/models.py [hunk 5]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+    votes = EmbeddedDocumentField(Votes)
+    created_at = DateTimeField()
+    updated_at = DateTimeField()
+    author_id = StringField()
+
+
+class CommentThread(Content):
+    """Represents a Thread."""
+    meta = {'collection': 'contents'}
+    _type = StringField(default='CommentThread')
```
Comment (jimabramson) [inline discussions/models.py:80]: readonly

File: discussions/models.py [hunk 6]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+    pinned = BooleanField()
+
+    @queryset_manager
+    def objects(doc_cls, queryset):
+        return queryset(_type='CommentThread')
+
+class Comment(Content):
+    """Represents a Comment."""
+    meta = {'collection': 'contents'}
+    _type = StringField(default='Comment')
```
Comment (jimabramson) [inline discussions/models.py:97]: readonly

File: discussions/models.py [hunk 7]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+        up_count = IntField(default=0)
+        down_count = IntField(default=0)
+        count = IntField(default=0)
+        point = IntField(default=0)
+
+    visible = BooleanField(default=True)
+    abuse_flaggers = ListField()
+    historical_abuse_flaggers = ListField()
+    author_username = StringField(default=None)
+    course_id = StringField()
```
Comment (jimabramson) [inline discussions/models.py]: readonly

File: discussions/models.py [hunk 8]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+        down = ListField()
+        up_count = IntField(default=0)
+        down_count = IntField(default=0)
+        count = IntField(default=0)
+        point = IntField(default=0)
+
+    visible = BooleanField(default=True)
+    abuse_flaggers = ListField()
+    historical_abuse_flaggers = ListField()
+    author_username = StringField(default=None)
```
Comment (jimabramson) [inline discussions/models.py]: readonly

File: discussions/models.py [hunk 9]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+    course_id = StringField()
+    body = StringField()
+    anonymous = BooleanField(default=False)
+    anonymous_to_peers = BooleanField(default=False)
+    at_position_list = ListField()
+    votes = EmbeddedDocumentField(Votes)
+    created_at = DateTimeField()
+    updated_at = DateTimeField()
+    author_id = StringField()
```
Comment (jimabramson) [inline discussions/models.py:74]: readonly

File: discussions/models.py [hunk 10]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+    @queryset_manager
+    def objects(doc_cls, queryset):
+        return queryset(_type='CommentThread')
+
+class Comment(Content):
+    """Represents a Comment."""
+    meta = {'collection': 'contents'}
+    _type = StringField(default='Comment')
+    endorsed = BooleanField(default=False)
+    endorsrement = DictField()
```
Comment (jimabramson) [inline discussions/models.py]: endorsement

File: discussions/models.py [hunk 11]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+    sk = StringField(default=None)
+    comment_thread_id = ObjectIdField()
```
Comment (jimabramson) [inline discussions/models.py]: readonly

File: discussions/models.py [hunk 12]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+    def objects(doc_cls, queryset):
+        return queryset(_type='CommentThread')
+
+class Comment(Content):
+    """Represents a Comment."""
+    meta = {'collection': 'contents'}
+    _type = StringField(default='Comment')
+    endorsed = BooleanField(default=False)
+    endorsrement = DictField()
+    sk = StringField(default=None)
```
Comment (jimabramson) [inline discussions/models.py]: readonly

File: discussions/models.py [hunk 13]
```diff
@@ -37,3 +42,65 @@ class ReadState(EmbeddedDocument):
...
+    comment_thread_id = ObjectIdField()
+    parent_ids = ListField()
```
Comment (jimabramson) [inline discussions/models.py]: readonly :-)

File: discussions/models.py (hunk 4 above)
Comment (dan-f) [inline discussions/models.py]: Got it.

File: discussions/models.py (hunk 5 above)
Comment (dan-f) [inline discussions/models.py:80]: I was thinking that we could continue to use the `_type` field for consistency. New models would need the field too in order to distinguish `Comment`s from `CommentThread`s in the collection.

File: discussions/models.py (hunk 7 above)
Comment (dan-f) [inline discussions/models.py]: How would we ever set this for new content?

File: discussions/models.py (hunk 8 above)
Comment (dan-f) [inline discussions/models.py]: Again, how do we set this for new documents?

File: discussions/models.py (hunk 10 above)
Comment (dan-f) [inline discussions/models.py]: got it

**Comment by dan-f** [comment]
@jimabramson read only/write once doesn't exist in mongoengine

**Comment by dan-f** [comment]
@jimabramson responded to review

**Comment by BenjiLee** [comment]
:+1:

**Comment by jimabramson** [comment]
:+1:

**Comment by tlindaliu** [comment]
:+1:

**Pull request merged**